chimedb/dataset/get.py

```diff
@@ -21,42 +24,281 @@
+                _logger.warning("Could not find dataset {}.".format(ds_id))
+                return None
+        return _dataset_cache[ds_id]
+
+    _type = None
+    _base_dataset = None
+
+    @property
+    def type(self):
+        """Get the type of the attached dataset state."""
+        if self._type is None:
+            return self.state.type
+        return self._type
+
+    @property
+    def base_dataset(self):
+        """Get the base dataset."""
+        if not self._base_dataset:
+            self._base_dataset = self.base_dset
+        return self.base_dset
+
+    @property
+    def dataset_state(self):
+        """Get the dataset state."""
+        # get the raw ID of the state (foreign key)
+        orm_base_class = super()
+        state_id = orm_base_class.state_id
+
+        if state_id not in _state_cache:
+            _state_cache[state_id] = orm_base_class.state
+        return _state_cache[state_id]
+
+    def __repr__(self):
+        if self._type:
+            return "<Dataset[{}]: {}>".format(self.type, self.id)
+        else:
+            return "<Dataset: {}>".format(self.id)
+
+    def closest_ancestor_of_type(self, type_):
+        """
+        Get the closest ancestorial dataset of the given type.
+
+        Parameters
+        ----------
+        type_ : str or :class:`chimedb.dataset.orm.DatasetStateType`
+
+        Raises
+        ------
+        :class:`chimedb.dataset.orm.DatasetStateType.DoesNotExist`
+            If `type_` is not an existing dataset state type.
+
+        Returns
+        -------
+        :class:`chimedb.dataset.orm.Dataset`
+            The closest ancestor of the given type. Returns `None` if no ancestor of the
+            given type found.
+        """
+        if isinstance(type_, str):
+            type_ = DatasetStateType.from_name(name=type_)
+            if type_ is None:
+                raise NotFoundError("{} is not a known DatasetStateType".format(type_))
+        d = self
+
+        while d:
+            if d.type == type_:
+                return d
+            d = d.base_dataset
+
+        return None
+
+
+def index():
+    """
+    Get an index of Datasets and the type of the DatasetStates they reference.
+
+    Returns
+    -------
+    dict
+        Index of Datasets and DatasetStateTypes
+    """
+    query = (
+        Dataset.select(Dataset, DatasetStateType.alias("_type"))
+        .join(DatasetState)
+        .join(DatasetStateType)
+    )
+
+    # Create a mapping from dataset ID to
+    dsdict = {d.id: d for d in query.objects(Dataset)}
+
+    # De-reference the base datasets
+    for d in dsdict.values():
+        base_dset = None if d.root or d.base_dset_id is None else dsdict[d.base_dset_id]
+        d._base_dataset = base_dset
+
+    # Pre-fetch all states without their .data: peewee caches this select and will not
+    # hit the database again
+    DatasetState.select(Dataset.id, DatasetState.time, DatasetState.type).execute()
```

Comment:
I'm pretty sure this exactly nothing. It only caches that data within the query object that can be iterated over.
```suggestion
```